File: app/src/main/java/seigneur/gauvain/mycourt/ui/pin/PinActivity.java
package seigneur.gauvain.mycourt.ui.pin;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.view.View;
import android.view.ViewStub;
import android.widget.Button;
import android.widget.Toast;

import com.alimuzaffar.lib.pin.PinEntryEditText;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import dagger.android.AndroidInjection;
import seigneur.gauvain.mycourt.R;
import seigneur.gauvain.mycourt.ui.base.BaseActivity;
import seigneur.gauvain.mycourt.utils.Constants;

public class PinActivity extends BaseActivity {

   /* @Inject
    PinPresenter mPinPresenter;*/

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    private PinViewModel mPinViewModel;

    @BindView(R.id.pinEditor)
    PinEntryEditText mPinEditor;

    @BindView(R.id.confirmPinBtn)
    Button mConfirmPinBtn;

    @BindView(R.id.stubPinChecker)
    ViewStub mStubPinChecker;
    //new instance of view stub to bind views with butterknife;
    ConfirmPinViewStub confirmPinViewStub;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AndroidInjection.inject(this);
        setContentView(R.layout.activity_pin);
        ButterKnife.bind(this);

        mPinViewModel = ViewModelProviders.of(this, viewModelFactory).get(PinViewModel.class);
        mPinViewModel.init();
        subscribeToLiveData(mPinViewModel);

        if (mPinEditor != null) {
            mPinEditor.setOnPinEnteredListener(str -> {
                mPinViewModel.onNewPinConfirmed(str.toString());
            });
        }

    }

    private void subscribeToLiveData(PinViewModel pinViewModel) {
        //observe current step
        pinViewModel.getStep().observe(this, this::setStepUI);
    }

    private void setStepUI(int step) {
        switch (step) {
            case Constants.PIN_STEP_CHECK_STORED :
                showConfirmCurrentPinView(true);
                break;
            case Constants.PIN_STEP_NEW_PIN_ONE :
                showConfirmCurrentPinView(false);
                break;
            case Constants.PIN_STEP_NEW_PIN_TWO :
                showConfirmCurrentPinView(false);
                break;
            default:
                showConfirmCurrentPinView(false);
        }
    }

    /**
     * user proposes string fro current pin code, check it in database
     * @param isVisible - check vsisiblity
     */
    public void showConfirmCurrentPinView(boolean isVisible) {
        if (isVisible) {
            if (mStubPinChecker.getParent() != null) {
                View inflated = mStubPinChecker.inflate();
                confirmPinViewStub = new ConfirmPinViewStub(inflated);
            } else {
                mStubPinChecker.setVisibility(View.VISIBLE);
            }
            //listen Pin entry
            confirmPinViewStub.currentPinInput.setOnPinEnteredListener(new PinEntryEditText.OnPinEnteredListener() {
                @Override
                public void onPinEntered(CharSequence str) {
                    mPinViewModel.onCurrentPinConfirmed(str.toString());
                }
            });

        }
        else {
            mStubPinChecker.setVisibility(View.GONE);
        }
    }

    /**
     * Inner class to bind views from ViewStub
     */
    public class ConfirmPinViewStub {
        @BindView(R.id.current_pin_checker)
        PinEntryEditText currentPinInput;
        public ConfirmPinViewStub(View view) {
            ButterKnife.bind(this, view);
        }
    }


}

File: app/src/main/java/seigneur/gauvain/mycourt/ui/pin/PinActivity.kt
package seigneur.gauvain.mycourt.ui.pin

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import android.os.Bundle
import android.view.View
import android.view.ViewStub
import android.widget.Button
import android.widget.Toast

import com.alimuzaffar.lib.pin.PinEntryEditText

import javax.inject.Inject

import butterknife.BindView
import butterknife.ButterKnife
import dagger.android.AndroidInjection
import seigneur.gauvain.mycourt.R
import seigneur.gauvain.mycourt.ui.base.BaseActivity
import seigneur.gauvain.mycourt.utils.Constants

class PinActivity : BaseActivity() {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private val mPinViewModel: PinViewModel by lazy {
        ViewModelProviders.of(this, viewModelFactory).get(PinViewModel::class.java)
    }

    @BindView(R.id.pinEditor)
    lateinit var mPinEditor: PinEntryEditText

    @BindView(R.id.confirmPinBtn)
    lateinit var mConfirmPinBtn: Button

    @BindView(R.id.stubPinChecker)
    lateinit var mStubPinChecker: ViewStub
    //new instance of view stub to bind views with butterknife;
    private lateinit var confirmPinViewStub: ConfirmPinViewStub


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AndroidInjection.inject(this)
        setContentView(R.layout.activity_pin)
        ButterKnife.bind(this)
        mPinViewModel.init()
        subscribeToLiveData(mPinViewModel)
        mPinEditor.setOnPinEnteredListener { str -> mPinViewModel.onNewPinConfirmed(str.toString()) }


    }

    private fun subscribeToLiveData(pinViewModel: PinViewModel) {
        //observe current step
        pinViewModel.step.observe(this, Observer<Int> { this.setStepUI(it!!) })
    }


    private fun setStepUI(step: Int) {
        when (step) {
            Constants.PIN_STEP_CHECK_STORED -> showConfirmCurrentPinView(true)
            Constants.PIN_STEP_NEW_PIN_ONE -> showConfirmCurrentPinView(false)
            Constants.PIN_STEP_NEW_PIN_TWO -> showConfirmCurrentPinView(false)
            else -> showConfirmCurrentPinView(false)
        }
    }

    /**
     * user proposes string fro current pin code, check it in database
     * @param isVisible - check vsisiblity
     */
    private fun showConfirmCurrentPinView(isVisible: Boolean) {
        if (isVisible) {
            if (mStubPinChecker.parent != null) {
                val inflated = mStubPinChecker.inflate()
                confirmPinViewStub = ConfirmPinViewStub(inflated)
            } else {
                mStubPinChecker.visibility = View.VISIBLE
            }
            //listen Pin entry
            confirmPinViewStub.currentPinInput.setOnPinEnteredListener { str -> mPinViewModel.onCurrentPinConfirmed(str.toString()) }

        } else {
            mStubPinChecker.visibility = View.GONE
        }
    }

    /**
     * Inner class to bind views from ViewStub
     */
    inner class ConfirmPinViewStub(view: View) {
        @BindView(R.id.current_pin_checker)
        lateinit var currentPinInput: PinEntryEditText

        init {
            ButterKnife.bind(this, view)
        }
    }


}
